### DISCUSSION TITLE
How To Get Discord Access Token

### DISCUSSION THREAD

**Discussion by timnikolsky** [nextauthjs/next-auth#2795] [Help] [answered]
Hi! I'm using Discord provider and I want to get Discord access token for their API. Where or how can I get it?

**Comment by balazsorban44** [comment, marked as answer]
https://next-auth.js.org/configuration/callbacks#jwt-callback

**Reply by timnikolsky** [reply]
Thank you! But I ran into a problem - my account is always undefined, so I never get token. Any Idea why?

**Reply by balazsorban44** [reply]
You only get it on the first call, as documented. Have another good read of the docs I linked to!:

> Use an if branch to check for the existence of parameters (apart from token). If they exist, this means that the callback is being invoked for the first time (i.e. the user is being signed in). This is a good place to persist additional data like an access_token in the JWT. Subsequent invocations will only contain the token parameter.

**Reply by timnikolsky** [reply]
> .

Yeah, I know that, but my account is undefined even on the first call. I tried to log out and log in back, but nothing happened

**Reply by balazsorban44** [reply]
That's sounds strange. Do you have a reproduction? Hard to guess. A public repository would be the most useful!

**Reply by timnikolsky** [reply]
Unfortunately cant do it now, but will show you the code tomorrow. Thanks for help!

**Reply by timnikolsky** [reply]
Cant show you the repository, but here is the code:
```ts
...
providers: [
        Providers.Discord({
            clientId: process.env.DISCORD_CLIENT_ID,
            clientSecret: process.env.DISCORD_CLIENT_SECRET,
            scope: "identify email guilds guilds.join",
        })
    ],
callbacks: {
	async session(session, user) {
		return session
	},
	async jwt(token, user, account = {}, profile, isNewUser) {
		// account is always an empty object (undefined)
		if (account.provider && !token[account.provider]) {
			token[account.provider] = {};
		}
		if (account.accessToken) {
			
			token[account.provider].accessToken = account.accessToken;
		}
		if (account.refreshToken) {
			
			token[account.provider].refreshToken = account.refreshToken;
		}
		return token;
	}
},

secret: process.env.NEXTAUTH_SECRET,
...
```

```ts
import type { NextApiRequest, NextApiResponse } from 'next'
import { getSession, useSession, getCsrfToken } from 'next-auth/client'
import { getToken } from 'next-auth/jwt'

export default async (req: NextApiRequest, res: NextApiResponse) => {
	const session = await getSession({ req })
	// Always null
	const token = await getToken({ req, secret: process.env.NEXTAUTH_SECRET })
	...
}

**Reply by timnikolsky** [reply]
@balazsorban44 Sorry for mention, but any idea how to fix it?

**Reply by balazsorban44** [reply]
I could not reproduce this locally, I get the `access_token` as expected, and `account` is not `undefined` for me on the initial jwt call.

```js
async jwt(token, user, account, profile, isNewUser) {
  if (account) {
    console.log(account)
  }
  return token
},
```
Logs this:
```js
{
  provider: 'discord',
  type: 'oauth',
  id: 'id',
  accessToken: 'some-access-token',
  accessTokenExpires: null,
  refreshToken: 'some-refresh-token',
  idToken: undefined,
  access_token: 'some-access-token',
  expires_in: 604800,
  refresh_token: 'some-refresh-token',
  scope: 'identify email',
  token_type: 'Bearer'
}
```

I'll either need a reproduction I can clone and run locally, or you have to keep digging because the error in your implementation somewhere.

Don't know if relevant, but your `jwt` callback snippet looks error-prone. You check for `account.provider`, but `account` is `undefined` on any subsequent requests, so you should rather do something like `account?.provider` ([optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)), or `if(account)` instead.

**Reply by timnikolsky** [reply]
I turn account into empty object by default, so i don't think the mistake is here. I'll open my repository today

**Reply by timnikolsky** [reply]
Here it is
https://github.com/sadnesszephyr/MonoWebsite/tree/dev

**Reply by balazsorban44** [reply]
Yep, works perfectly. Here is what `account` looks like for me:
```js
{
  provider: 'discord',
  type: 'oauth',
  id: 'id',
  accessToken: 'access-token',
  accessTokenExpires: null,
  refreshToken: 'refresh-token',
  idToken: undefined,
  access_token: 'access-token',
  expires_in: 604800,
  refresh_token: 'refresh-token',
  scope: 'identify email guilds guilds.join',
  token_type: 'Bearer'
}
```

You should revisit the if statements in your `jwt` callback, and also check the `session` callback to forward it to the client (if that's something you need)

The docs for v3: https://next-auth.js.org/v3/configuration/callbacks

**Reply by timnikolsky** [reply]
Wow, i didn't change anything and it strated working. As usual. Huge thank you :)
